common/persistence/pinotUserQueryTimeFilter.go

```diff
@@ -0,0 +1,170 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package persistence
+
+import (
+	"errors"
+	"fmt"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/definition"
+	"github.com/uber/cadence/common/types"
+	"github.com/xwb1989/sqlparser"
+	"strings"
+)
+
+// VisibilityQueryFilter for sql query validation
+type VisibilityQueryFilter struct {
+}
+
+// NewVisibilityQueryFilter create VisibilityQueryValidator
+func NewVisibilityQueryFilter() *VisibilityQueryFilter {
+	return &VisibilityQueryFilter{}
+}
+
+// FilterQuery validates that search attributes in the query and returns modified query.
+func (qv *VisibilityQueryFilter) FilterQuery(whereClause string) (string, error) {
+	if len(whereClause) != 0 {
+		// Build a placeholder query that allows us to easily parse the contents of the where clause.
+		// IMPORTANT: This query is never executed, it is just used to parse and validate whereClause
+		var placeholderQuery string
+		whereClause := strings.TrimSpace(whereClause)
+		if common.IsJustOrderByClause(whereClause) { // just order by
+			placeholderQuery = fmt.Sprintf("SELECT * FROM dummy %s", whereClause)
+		} else {
+			placeholderQuery = fmt.Sprintf("SELECT * FROM dummy WHERE %s", whereClause)
+		}
+
+		stmt, err := sqlparser.Parse(placeholderQuery)
+		if err != nil {
+			return "", &types.BadRequestError{Message: "Invalid query."}
+		}
+
+		sel, ok := stmt.(*sqlparser.Select)
+		if !ok {
+			return "", &types.BadRequestError{Message: "Invalid select query."}
+		}
+		buf := sqlparser.NewTrackedBuffer(nil)
+		res := ""
+		// validate where expr
+		if sel.Where != nil {
+			res, err = qv.validateWhereExpr(sel.Where.Expr)
+			if err != nil {
+				return "", &types.BadRequestError{Message: err.Error()}
+			}
+		}
+
+		sel.OrderBy.Format(buf)
+		res += buf.String()
+		return res, nil
+	}
+	return whereClause, nil
+}
+
+func (qv *VisibilityQueryFilter) validateWhereExpr(expr sqlparser.Expr) (string, error) {
+	if expr == nil {
+		return "", nil
+	}
+	switch expr := expr.(type) {
+	case *sqlparser.AndExpr, *sqlparser.OrExpr:
+		return qv.filterAndOrExpr(expr)
+	case *sqlparser.ComparisonExpr:
+		return qv.filterComparisonExpr(expr)
+	case *sqlparser.RangeCond:
+		return qv.filterRangeExpr(expr)
+	case *sqlparser.ParenExpr:
+		return qv.validateWhereExpr(expr.Expr)
+	default:
+		return "", errors.New("invalid where clause")
+	}
+}
+
+// for "between...and..." only
+func (qv *VisibilityQueryFilter) filterRangeExpr(expr sqlparser.Expr) (string, error) {
+	buf := sqlparser.NewTrackedBuffer(nil)
+	rangeCond := expr.(*sqlparser.RangeCond)
+	colName, ok := rangeCond.Left.(*sqlparser.ColName)
+	if !ok {
+		return "", errors.New("invalid range expression: fail to get colname")
+	}
+	colNameStr := colName.Name.String()
+
+	if colNameStr == definition.StartTime {
```

Comment:
You filtered the startTime query here, is it because startTIme will be defined in the comparator workflow?

Reply:
yes

Reply:
What if user passed CloseTime=missing, since we have list open conditions in comparator workflow.